### PR TITLE
Update `resume_throw`

### PR DESCRIPTION
**Pull request by dhil**
This PR updates `resume_throw` such that it is parameterised by both an exception and an optional filter list.

New tests still need to be added to cover the extended syntax and semantics.

### CODE REVIEW

File: interpreter/binary/encode.ml [hunk 1]
```diff
@@ -233,7 +233,7 @@ struct
     | ContBind x -> op 0xe1; var x
     | Suspend x -> op 0xe2; var x
     | Resume xls -> op 0xe3; vec var_pair xls
-    | ResumeThrow x -> op 0xe4; var x
+    | ResumeThrow (x,xls) -> op 0xe4; var x; vec var_pair xls
```
Comment (rossberg) [inline interpreter/binary/encode.ml]:
```suggestion
    | ResumeThrow (x, xls) -> op 0xe4; var x; vec var_pair xls
```

File: interpreter/valid/valid.ml [hunk 2]
```diff
@@ -155,8 +155,6 @@ let check_def_type (c : context) (dt : def_type) at =
   | DefFuncT ft -> check_func_type c ft at
   | DefContT ct -> check_cont_type c ct at
 
-
-
 (* Stack typing *)
```
Comment (rossberg) [inline interpreter/valid/valid.ml:159]:
```suggestion

(* Stack typing *)
```
Nit: I use the convention to have 2 empty lines before "section header comments" like this.

File: interpreter/valid/valid.ml [hunk 3]
```diff
@@ -210,7 +208,6 @@ let peek_ref i (ell, ts) at : ref_type =
       ("type mismatch: instruction requires reference type" ^
        " but stack has " ^ string_of_val_type t)
 
-
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Same here.

File: interpreter/valid/valid.ml [hunk 4]
```diff
@@ -325,6 +322,24 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
...
+         let ContT z' = cont_type c (y' @@ x2.at) in
+         let ft' = func_type c (as_stat_var z' @@ x2.at) in
+         require (match_func_type c.types (FuncT (ts4, ts2)) ft') x2.at
+           "type mismatch in continuation type";
+         check_stack c (ts3 @ [RefT (nul', DefHT (Stat y'))]) ts' x2.at
+      | _ ->
+         error at
+           ("type mismatch: instruction requires continuation reference type" ^
+              " but label has " ^ string_of_result_type ts')
+    ) xys
```
Comment (rossberg) [inline interpreter/valid/valid.ml]:
```suggestion
    let TagT x1' = tag c x1 in
    let FuncT (ts3, ts4) = func_type c (as_stat_var x1' @@ x1.at) in
    let (_, ts') = label c x2 in
    match Lib.List.last_opt ts'  with
    | Some (RefT (nul', DefHT (Stat y'))) ->
      let ContT z' = cont_type c (y' @@ x2.at) in
      let ft' = func_type c (as_stat_var z' @@ x2.at) in
      require (match_func_type c.types (FuncT (ts4, ts2)) ft') x2.at
        "type mismatch in continuation type";
      check_stack c (ts3 @ [RefT (nul', DefHT (Stat y'))]) ts' x2.at
    | _ ->
      error at
        ("type mismatch: instruction requires continuation reference type" ^
          " but label has " ^ string_of_result_type ts')
  ) xys
```

**Review by rossberg — APPROVED** [review]

**Comment by dhil** [comment]
Thank you for the review!

**Pull request merged**